rock-paper-scissors/src/main/kotlin/rps/project/Game.kt

```diff
@@ -0,0 +1,85 @@
+package rps.project
+
+enum class PlayerName {
+    PLAYER1, PLAYER2
+}
+
+enum class MoveName {
+    ROCK, PAPER, SCISSORS, END_GAME
+}
+
+class Game(private val p1: Player, private val p2: Player, actions: Set<Action>) {
+
+    private val actionMap = actions.associateBy({ a -> a.name }, { a -> a } )
+
+    // Not strictly necessary, but it cleans up some of the code to use an empty action as a starting state
+    private val emptyAction = Action(null, setOf())
+
+    private var pipeline = { _: PlayerName, _: Action -> Triple(PlayerName.PLAYER1, emptyAction, Record(0,0,0)) }
+
+    private fun appendNewActionToPipeline(action: Action, callback: (Action, Action, Outcome) -> Unit) {
+        val curriedFunction = { currentPlayer: PlayerName, lastAction: Action -> processAction(currentPlayer, lastAction, action, callback) }
+        this.pipeline = bind(this.pipeline, curriedFunction)
+    }
+
```

Comment:
The rest of the code was fairly straightforward, but this is where I had a little bit of fun with functional programming patterns and did it in such a way that some of these functions (marked below) could be made asynchronous without really changing anything. For example, if we wanted to run in a browser and wait for Player 1 and Player 2 to make their moves without blocking the main thread.

For the sake of keeping the code simple and readable though, it is fully synchronous.